src/fiber/fiber.ml

```diff
@@ -208,12 +222,47 @@ end = struct
       forward_error exn;
       run_jobs jobs
 
-  let run fiber ~iter =
+  let with_context t ~f =
     let backup = !current in
     Exn.protect
       ~finally:(fun () -> current := backup)
       ~f:(fun () ->
-        let t = create () in
+        current := t;
+        f ())
+
+  module Scheduler = struct
+    type 'a state =
+      { mutable result : 'a option
+      ; context : t
+      }
+
+    type 'a stalled = 'a state
+
+    type 'a step =
+      | Done of 'a
+      | Stalled of 'a stalled
+
+    let start fiber =
+      let context = create () in
+      let state = { result = None; context } in
+      with_context context ~f:(fun () ->
+          apply (fun () -> fiber) () (fun x -> state.result <- Some x);
+          run_jobs context.jobs;
+          match state.result with
+          | Some x -> Done x
+          | None -> Stalled state)
+
+    let advance state =
+      with_context state.context ~f:(fun () ->
+          run_jobs state.context.jobs;
+          match state.result with
+          | Some x -> Done x
+          | None -> Stalled state)
```

Comment:
Everything starting from run_jobs seems is duplicated in start and advance. I'd factor that out if possible.

Reply:
Sounds good, done